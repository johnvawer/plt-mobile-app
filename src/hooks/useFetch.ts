import { useEffect, useState } from 'react'
import axios, { AxiosResponse } from 'axios'

interface ReturnPayload {
  data: AxiosResponse<any> | null
  loading: boolean
  error: boolean
}

const baseUrl = 'https://my-json-server.typicode.com/benirvingplt'

export const useFetch = (url: string): ReturnPayload => {
  const [loading, setLoading] = useState<boolean>(true)
  const [data, setData] = useState<AxiosResponse<any> | null>(null)
  const [error, setError] = useState<boolean>(false)

  useEffect(() => {
    if (url === '') return

    const fetchData = (): void => {
      setLoading(true)
      axios.get(`${baseUrl}${url}`)
        .then((response: AxiosResponse<any>) => {
          setData(response)
        })
        .catch(e => {
          console.log(e)
          setError(true)
        })
        .finally(() => {
          setLoading(false)
        })
    }

    fetchData()
  }, [url])

  return { loading, data, error }
}
