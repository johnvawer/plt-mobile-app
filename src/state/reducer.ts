type ActionMap<M extends { [index: string]: any }> = {
  [Key in keyof M]: M[Key] extends undefined
    ? {
      type: Key
    }
    : {
      type: Key
      payload: M[Key]
    }
}

export enum Types {
  Add = 'ADD_PRODUCT',
  Remove = 'REMOVE_PRODUCT'
}

interface Product {
  id: number
  name: string
  price: number
  quantity: number
  totalPrice: number
}

interface ProductPayload {
  [Types.Add]: {
    id: number
    name: string
    price: number
  }
  [Types.Remove]: {
    id: number
  }
}

export type ProductActions = ActionMap<ProductPayload>[keyof ActionMap<ProductPayload>]

export const productReducer = (state: Product[], action: ProductActions): Product[] => {
  const itemExists = state.find(product => product.id === action.payload.id)

  switch (action.type) {
    case 'ADD_PRODUCT':
      if (itemExists != null) {
        return state.map(product =>
          product.id === itemExists.id
            ? {
              ...product,
              quantity: product.quantity + 1,
              totalPrice: product.price * (product.quantity + 1)
            } : product
        )
      }

      return [
        ...state,
        {
          id: action.payload.id,
          name: action.payload.name,
          price: action.payload.price,
          quantity: 1,
          totalPrice: action.payload.price
        }
      ]
    case 'REMOVE_PRODUCT':
      if ((itemExists != null) && itemExists.quantity > 1) {
        return state.map(product =>
          product.id === itemExists.id
            ? {
              ...product,
              quantity: product.quantity - 1,
              totalPrice: product.price * (product.quantity - 1)
            } : product
        )
      }

      return [
        ...state.filter(product => product.id !== action.payload.id)
      ]
    default:
      return state
  }
}
