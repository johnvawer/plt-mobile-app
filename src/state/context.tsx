import { createContext, Dispatch } from 'react'
import { productReducer, ProductActions } from './reducer'
interface ComputedProduct {
  id: number
  name: string
  price: number
  quantity: number
  totalPrice: number
}

interface InitialState {
  products: ComputedProduct[]
}

export const initialState = {
  products: []
}

export const AppContext = createContext<{
  state: InitialState
  dispatch: Dispatch<ProductActions>
}>({
  state: initialState,
  dispatch: () => null
})

export const mainReducer = ({ products }: InitialState, action: ProductActions): { products: ComputedProduct[] } => ({
  products: productReducer(products, action)
})
