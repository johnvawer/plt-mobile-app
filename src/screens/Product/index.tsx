import React, { useContext } from 'react'
import { View, Text, StyleSheet } from 'react-native'
import { RouteProp } from '@react-navigation/native'
import { HomeStackParamsList } from '../../types/routes'

import Loading from '../../components/Atoms/Loading'
import ProductCard from '../../components/Molecules/ProductCard'

import { Types } from '../../state/reducer'
import { useFetch } from '../../hooks/useFetch'
import { AppContext } from '../../state/context'

interface Props {
  route: RouteProp<HomeStackParamsList, 'Product'>
}

const Product = ({ route }: Props): JSX.Element => {
  const { productId } = route.params
  const { loading, data, error } = useFetch(`/products/products/${productId}`)
  const { dispatch } = useContext(AppContext)

  if (loading) return <Loading />

  if (error || data === null) return <Text>Error fetching product</Text>

  const product: Product = data.data

  const onPress = (): void => {
    const { id, name, price } = product

    dispatch({
      type: Types.Add,
      payload: {
        id,
        name,
        price
      }
    })
  }

  return (
    <View style={styles.container}>
      <ProductCard
        onPress={onPress}
        product={product}
        buttonText='Add Product to Basket'
      />
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'stretch'
  }
})

export default Product
