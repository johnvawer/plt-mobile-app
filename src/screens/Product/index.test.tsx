import React from 'react'
import { render, fireEvent } from '@testing-library/react-native'
import * as fetchHook from '../../hooks/useFetch'
import { AppContext } from '../../state/context'
import Product from '.'

const fakeProduct = {
  id: 1,
  name: 'product name',
  colour: 'black',
  price: 10,
  img: 'http://someimage.png',
  quantity: 1,
  totalPrice: 1000
}

describe('<Product />', () => {
  let useFetchMock: jest.SpyInstance
  let dispatchMock: jest.Mock
  let initialState: {
    state: { products: [] }
    dispatch: () => void
  }
  let route: any

  beforeEach(() => {
    useFetchMock = jest.spyOn(fetchHook, 'useFetch')
    dispatchMock = jest.fn()
    initialState = {
      state: { products: [] },
      dispatch: dispatchMock
    }
    route = {
      key: 'key',
      name: 'Product',
      params: { productId: 1 },
      path: undefined
    }
  })

  afterEach(() => {
    useFetchMock.mockRestore()
    dispatchMock.mockRestore()
  })

  it('should call to fetch the product from the api', () => {
    useFetchMock.mockReturnValueOnce({
      loading: false,
      data: {
        data: fakeProduct
      },
      error: false
    })
    render(
      <AppContext.Provider value={initialState}>
        <Product route={route} />
      </AppContext.Provider>
    )

    expect(useFetchMock).toHaveBeenCalledTimes(1)
    expect(useFetchMock).toHaveBeenCalledWith('/products/products/1')
  })

  it('should display a loading message when the data is loading', () => {
    useFetchMock.mockReturnValueOnce({
      loading: true,
      data: null,
      error: false
    })
    const { toJSON } = render(
      <AppContext.Provider value={initialState}>
        <Product route={route} />
      </AppContext.Provider>
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should display an error message when an error is returned', () => {
    useFetchMock.mockReturnValueOnce({
      loading: false,
      data: null,
      error: true
    })
    const { toJSON } = render(
      <AppContext.Provider value={initialState}>
        <Product route={route} />
      </AppContext.Provider>
    )

    expect(toJSON()).toMatchSnapshot()
  })

  it('should dispatch a product to the store on button press', () => {
    useFetchMock.mockReturnValueOnce({
      loading: false,
      data: {
        data: fakeProduct
      },
      error: false
    })
    const { getByText } = render(
      <AppContext.Provider value={initialState}>
        <Product route={route} />
      </AppContext.Provider>
    )
    const addProductButton = getByText('Add Product to Basket')
    fireEvent.press(addProductButton)

    expect(dispatchMock).toHaveBeenCalledTimes(1)
    expect(dispatchMock).toHaveBeenCalledWith({
      type: 'ADD_PRODUCT',
      payload: {
        id: 1,
        name: 'product name',
        price: 10
      }
    })
  })
})
