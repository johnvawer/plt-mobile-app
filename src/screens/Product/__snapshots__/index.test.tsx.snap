// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> should display a loading message when the data is loading 1`] = `
<View>
  <Text>
    Loading
  </Text>
</View>
`;

exports[`<Product /> should display an error message when an error is returned 1`] = `
<Text>
  Error fetching product
</Text>
`;
